kvbc/include/categorization/block_merkle_category.h

```diff
@@ -70,20 +70,107 @@ class BlockMerkleCategory {
+  // new root hash is calculated for all remaining active keys in the block, and a new version of
+  // the `MerkleBlockValue` is written into the block merkle tree. Lastly, each active key hash is
+  // written into a separate column family with the pruned block_id as the value. These latter
+  // active keys are necessary for garbage collection, while the `PrunedBlock` and `MerkleBlockValue`
+  // are used to provide (and verify) proofs for active keys in pruned blocks.
+  //
+  // Garbage collection occurs when the block that first updated or deleted an active key from a
+  // prior prune block is itself pruned. All keys in the current block being pruned are checked
+  // during pruning to see if there were prior active keys from pruned blocks. If so, those prior
+  // versions of the keys are deleted along with their active references used for garbage
+  // collection. If any active keys remain for a pruned block, a new root hash is calculated, a new
+  // PrunedBlock written, and the merkle tree is updated yet again for that block. If no active keys
+  // remain for that pruned block, then the pruned block is deleted and the block is finally removed
+  // from the merkle tree, as there are no longer any keys from that block eligible for proofs. This
+  // garbage collection and update process is what is performed by this function.
+  //
+  // To reiterate: pruning (calling `deleteGenesisBlock`) does not always fully delete all data in a
+  // block. If a key is at the latest version when the block is deleted, it is still `active` and
+  // users can retrieve its data and prove that it is part of the blockchain. Any active keys become
+  // `deactivated` when a new version of the key is written in a later block, or the key is deleted
+  // in a later block. However, even though the key is no longer the latest version, it is not
+  // deleted in this step. Final deletion from the database only occurs when the overwriting block
+  // is itself pruned. This tradeoff is made to defer work to the pruning process and not cause
+  // block addition to slow down.
+  void rewriteAlreadyPrunedBlocks(std::unordered_map<BlockId, std::vector<KeyHash>>& deleted_keys,
+                                  storage::rocksdb::NativeWriteBatch& batch);
+
+  // When a genesis block is pruned, we must delete all data that is stale as of `tree_version`.
+  //
+  // This includes data that may be written in prior tree versions as a result of prior pruning
+  // operations that generate new tree versions and stale data, but not new blocks. There can be a
+  // large amount of these prior versions, as each pruned block generates a new tree version. If we
+  // prune X blocks in a row before we add a new block, Y, then when we prune Block Y, we will have
+  // to lookup all the indexes for those pruned X tree versions and delete the internal and leaf
+  // nodes in the indexes.
```

Comment:
Nit: Maybe we can call the number of blocks we prune in a row `N` and the newly-added block `X`.